File: leetcode/algorithms/685-RedundantConnectionII/findRedundantDirectedConnection.cpp
class Solution {
public:
    /**
     * Time O(n)
     * Space O(n)
     */
    vector<int> findRedundantDirectedConnection(vector<vector<int>>& edges) {
        // Build graph
        Graph g;
        g.buildGraph(edges.size(), edges);
        
        // Case 1: One vertex has two parents
        // Implementation: graph traversal
        // O(N), where N is the number of nodes
        if(g.candidatesSize() > 0)
            return g.redundantCandidate();
        
        // Case 2: Every vertex has one parent (cycle)
        // Implementation: loop over edges
        // O(N), where N is the number of edges
        return g.lastCycleEdge(edges);
    }
    
    
private:
    struct Node {
        int val;
        unordered_set<int> adj; // used unordered_set for quick insert/remove element
        vector<int> parentEdge; // track the parent node
        
        Node(int v=0):val(v),adj(){}
    };
    
    class Graph {
        vector<Node> nodes;
        unordered_set<int> visited;
        vector<vector<int>> candidates;
        
    public:
        void buildGraph(int size, vector<vector<int>> edges) {
        
            for(int i=0; i<=size; i++) {
                Node n;
                n.val = i;
                nodes.push_back(n);
            }
            
            for(int i=0; i<size; i++) {
                int u = edges[i][0];
                int v = edges[i][1];
                if (nodes[v].parentEdge.size() == 0) { // first parent
                    nodes[v].parentEdge = edges[i];
                    addEdge(u, v); // only add first edge, test second edge
                }
                else { // EITHER one node only with two parents OR no node with two parents
                    candidates.push_back(nodes[v].parentEdge); // last edge
                    candidates.push_back(edges[i]); // current edge
                }
            }
                
        }
 
        void addEdge(int u, int v) {
            nodes[u].adj.insert(v);
        }

        void removeEdge(int u, int v) {
            nodes[u].adj.erase(v);
        }
        
        int candidatesSize() {
            return candidates.size();
        }
        
        vector<int> lastCycleEdge(vector<vector<int>>& edges) {
            visited.clear();
            vector<int> cycleEdge;
            
            // find last cycle
            for(int i=0; i<edges.size(); i++) {
                // Check cycle
                if(visited.count(edges[i][1])>0) {
                    // found cycle
                    cycleEdge = edges[i];
                }
                
                // Mark vertices visited
                visited.insert(edges[i][0]);
                visited.insert(edges[i][1]);
            }
            
            // Return last cycle eedge
            return cycleEdge;
        }
        
        vector<int> redundantCandidate() {
            // Find root
            int root;
            for (root=1; root<nodes.size(); root++) {
                if (nodes[root].parentEdge.size() == 0) {
                    break;
                }
            }
            
            // Test graph connectivity (with one candidate edge removed)
            if (dfsTraversal(nodes[root].val))
                    return candidates[1]; // visited all nodes
            
            // The redundant edge must be the other candidate
            return candidates[0];
        }
        
        bool dfsTraversal(int u) {
            // Mark node visited
            visited.insert(u); 
            
            // Base case: found all nodes in graph
            if (visited.size() == nodes.size()-1) // Ignore 0th element in nodes
                return true;
            
            // Recursive depth first search
            for(auto it=nodes[u].adj.begin(); it!=nodes[u].adj.end(); it++) {
                // Check if visited
                if(visited.count(*it)>0)
                    continue; // skip
                
                // Recursive search
                if (dfsTraversal(*it))
                    return true;
            }
            
            // not all nodes visited
            return false;   
        }
    }; 
};

File: leetcode/algorithms/685-RedundantConnectionII/findRedundantDirectedConnection.java
class Solution {
    // This class represents a directed graph using adjacency list
    // representation
    class Graph
    {
        private int V;   // No. of vertices
        private LinkedList<Integer> adj[]; //Adjacency Lists
        private int[] numConnections; // num of connections pointing to node

        // Constructor
        Graph(int v)
        {
            V = v;
            adj = new LinkedList[v];
            for (int i=0; i<v; ++i)
                adj[i] = new LinkedList();
            numConnections = new int[v];
        }

        // Function to add an edge into the graph
        void addEdge(int v,int w)
        {
            adj[v].add(w);
            numConnections[w]++;
        }
        
        // Function to remove an edge into the graph
        void removeEdge(int v,int w)
        {
            adj[v].removeFirstOccurrence(w);
            numConnections[w]--;
        }
        
        // Funciton to return number of connections directed towards node
        int getNumConnections(int v) {
            return numConnections[v];
        }

        // prints BFS traversal from a given source s
        boolean BFS(int s)
        {
            // Mark all the vertices as not visited(By default
            // set as false)
            boolean visited[] = new boolean[V];

            // Create a queue for BFS
            LinkedList<Integer> queue = new LinkedList<Integer>();

            // Mark the current node as visited and enqueue it
            visited[s]=true;
            queue.add(s);

            while (queue.size() != 0)
            {
                // Dequeue a vertex from queue and print it
                s = queue.poll();
                //System.out.print(s+" ");

                // Get all adjacent vertices of the dequeued vertex s
                // If a adjacent has not been visited, then mark it
                // visited and enqueue it
                Iterator<Integer> i = adj[s].listIterator();
                while (i.hasNext())
                {
                    int n = i.next();
                    if (!visited[n])
                    {
                        visited[n] = true;
                        queue.add(n);
                    }
                }
            }
            
            for (int i=1; i<V; i++) {
                if (!visited[i]) {// missing node(s)
                    //System.out.println("Missing node " + i);
                    return false;
                }
            }
            
            return true; // all nodes visited (connected tree)
        }
    }
    // This code is contributed by Aakash Hasija
    
    public int[] findRedundantDirectedConnection(int[][] edges) {       
        Graph g = new Graph(edges.length+1);

        // add edges
        for (int i=0; i<edges.length; i++) {
            g.addEdge(edges[i][0], edges[i][1]);
        }

        for (int i=edges.length-1; i>=0; i--) {
            //System.out.println("Removed edge " + i);
            if(g.getNumConnections(edges[i][0]) == 0 && g.getNumConnections(edges[i][1]) == 1) 
                continue; // skip nodes with only one connection
            g.removeEdge(edges[i][0], edges[i][1]);
            for (int j=1; j<edges.length+1; j++) {
                //System.out.println("BFS on " + j);
                if(g.getNumConnections(j) > 0)
                    continue; // Skip if not root
                if(g.BFS(j))
                    return edges[i];
            }
            g.addEdge(edges[i][0], edges[i][1]);
            //System.out.println();
        }
        
        return edges[0];
    }
}
